File: lowest-common-ancestor-of-a-binary-tree/lowest-common-ancestor-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#Time: O(N)
#Space: O(N)
class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        
        def backtrack(path,root,v):
            #base
            if root is None:
                return
            if root==v:
                path.append(root)
            
            path.append(root)
            backtrack(path,root.left,v)

File: lowest-common-ancestor-of-a-binary-tree/lowest-common-ancestor-of-a-binary-tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode LowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        IList<TreeNode> plist = new List<TreeNode>();
        IList<TreeNode> qlist = new List<TreeNode>();
        Helper(root,p,plist);
        Helper(root,q,qlist);
        
        TreeNode res = null;
        int i=0;
        int j=0;
        
        while(true)
        {
            if(i >=plist.Count)
            {
                break;
            }
            if(j >=qlist.Count)
            {
                break;
            }
            if(plist[i] != qlist[j])
            {
                break;
            }
            
            res = plist[i];
            i++;
            j++;
        }
        
        return res;
        
    }
    
    private void Helper(TreeNode node,TreeNode target,IList<TreeNode> path)
    {
        if(node ==null)
        {
            return;
        }
        
        path.Add(node);
        
        if(node==target)
        {
            return;
        }
        Helper(node.left,target,path);
        if(path[path.Count-1]==target)
        {
            return;
        }
        Helper(node.right,target,path);
        
        if(path[path.Count-1] == target)
        {
            return;
        }
        
        path.RemoveAt(path.Count-1);
    }
}
